kotlin/flatbuffers-kotlin/src/commonMain/kotlin/com/google/flatbuffers/kotlin/JSON.kt

```diff
@@ -0,0 +1,828 @@
+      // the remaining bytes
+      while (true) {
+        if (cursorPos >= limit) {
+          error("Unexpected end of string")
+        }
+        currentChar = fetch(cursorPos)
+        if (!isPlainStringChar(currentChar)) {
+          break
+        }
+        ++cursorPos
+      }
+    }
+    if (currentChar == CHAR_DOUBLE_QUOTE) {
+      val str = data.getString(readPos, cursorPos - readPos)
+      readPos = cursorPos + 1
+      return str
+    }
+    if (currentChar in 0..0x1f) {
+      error("Illegal Codepoint")
+    } else {
+      // backslash or >0x7f
+      return readStringSlow(data, currentChar, cursorPos)
+    }
+  }
+
+  private fun readStringSlow(data: ReadBuffer, first: Byte, lastPos: Int): String {
+    var cursorPos = lastPos
+
+    var endOfString = lastPos
+    while (true) {
+      val pos = data.findFirst(CHAR_DOUBLE_QUOTE, endOfString)
+      when {
+        pos == -1 -> makeError(data, "Unexpected EOF, missing end of string '\"'", first)
+        data[pos - 1] == CHAR_BACKSLASH && data[pos - 2] != CHAR_BACKSLASH -> {
+          // here we are checking for double quotes preceded by backslash. eg \"
+          // we have to look past pos -2 to make sure that the backlash is not
+          // part of a previous escape, eg "\\"
+          endOfString = pos + 1
+        }
+        else -> {
+          endOfString = pos; break
+        }
+      }
+    }
+    // copy everything before the escape
+    val builder = StringBuilder(data.getString(readPos, lastPos - readPos))
+    while (true) {
+      when (val pos = data.findFirst(CHAR_BACKSLASH, cursorPos, endOfString)) {
+        -1 -> {
+          val doubleQuotePos = data.findFirst(CHAR_DOUBLE_QUOTE, cursorPos)
+          if (doubleQuotePos == -1) makeError(data, "Reached EOF before enclosing string", first)
+          val rest = data.getString(cursorPos, doubleQuotePos - cursorPos)
+          builder.append(rest)
+          readPos = doubleQuotePos + 1
+          return builder.toString()
+        }
+
+        else -> {
+          // we write everything up to \
+          builder.append(data.getString(cursorPos, pos - cursorPos))
+          val c = data[pos + 1]
+          builder.append(readEscapedChar(data, c, pos))
+          cursorPos = pos + if (c == CHAR_u) 6 else 2
+        }
+      }
+    }
+  }
+
+  private inline fun isPlainStringChar(c: Byte): Boolean {
+    val flags = parseFlags
+    // return c in 0x20..0x7f && c != 0x22.toByte() && c != 0x5c.toByte()
+    return (flags[c.toInt() and 0xFF] and 1) != 0.toByte()
+  }
+
+  private inline fun isWhitespace(c: Byte): Boolean {
+    val flags = parseFlags
+    // return c == '\r'.toByte() || c == '\n'.toByte() || c == '\t'.toByte() || c == ' '.toByte()
+    return (flags[c.toInt() and 0xFF] and 2) != 0.toByte()
+  }
+
+  private fun reset() {
+    readPos = 0
+    output.clear()
+    scopes.reset()
+  }
+
+  private fun nextToken(data: ReadBuffer): Token {
+    val scope = this.scopes.last
```

Comment:
This is still very non-standard. What are all these scopes needed for? It still calls `skipWhitespace` all over the place. Please have a look at this function: https://github.com/google/flatbuffers/blob/276b1bc342d23142e4b2b9b9fadbf076474deec9/src/idl_parser.cpp#L333
Note it only needs to deal with whitespace in one location, and it isn't even a function. Also see how `Next` doesn't return a value, code actually becomes simpler with the concept of a current token instead of passing it around explicitly. Structuring your lexical analysis like this will make your code simpler, more robust, produces more logical errors and is probably faster. Also see here how this function is used to actually parse JSON: https://github.com/google/flatbuffers/blob/276b1bc342d23142e4b2b9b9fadbf076474deec9/src/idl_parser.cpp#L3021 Notice the use of EXPECT which takes care of the bulk of error handling.

Reply:
Hi @aardappel. I am not exactly sure how standard json parser should look like, but my idea was to follow closely what is considered the industry "standard" in term of JSON on JVM, which is the [Gson](https://github.com/google/gson) implementation.

You will see that `nextToken()` is basically the same logic of:

https://github.com/google/gson/blob/ceae88bd6667f4263bbe02e6b3710b8a683906a2/gson/src/main/java/com/google/gson/stream/JsonReader.java#L460

Another very used library, Moshi, also does very similar:

https://github.com/square/moshi/blob/7ef4a32d5224623028bfd4157856fc7960a08227/moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java#L243

> It still calls skipWhitespace all over the place

Actually its called only on 2 places: On `nextToken()` and at the `parse()`. On `parse()` the sole purpose is to make sure we catch extraneous characters after 
 the payload, like ```"[1,2,3] ,"```. There is a third place `parseNumber()` which actually is not correct and I am removing now.

Actually, if you may have a feeling this implementation is more complex (like the use of scopes) because it tries to be much more adherent to the RFC than the cpp implementation. For example, cpp is lenient with the extra characters at the end that I mentioned before:
```bash
$ echo "[1,2,3] ," > error.json
$ ./flatc --binary --flexbuffers error.json.  # This should fail
```
It also seems to accept a variety of other non-conformant payloads, like : `"[\"Extra close\"]]"`, `"{\"Extra comma\": true,}"`, `"{\"Numbers cannot have leading zeroes\": 013}"`

So in fact, we are dealing with extra space in one location only (the other is just for error checking). The `Scope` is simply a state that allows `nextToken` to parse correctly.  

As for having current token being stored as a field instead being returned and handle locally, I am not sure how this will beneficial, sounds more like a matter of style. Why do you think this will have performance benefits? I can run the benchmarks, but I feel it might be pointless.

Reply:
I am simply stating how a lexical analyzer is typically written (in compilers, anyway), and why it leads to better code, I have no idea why Gson and friends do it the way they do.

If there is a bug in the C++ parser, we can fix that separately.

Either way, sounds like you prefer to follow what these implementations do, so be it.